### PR TITLE
side effect expressions with operands

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 662434d11459133b37696a47adeeb773a6c43c6a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97953188

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 24c0de437252917ecaf01f4f384636a9517522fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98077159

File: src/util/std_code.h [hunk 1]
```diff
@@ -1557,18 +1557,19 @@ inline const code_expressiont &to_code_expression(const codet &code)
 /// subtypes are not subtypes of \ref codet, but they inherit directly from
 /// \ref exprt. They do have a `statement` like [codets](\ref codet), but their
 /// [id()](\ref irept::id) is `ID_side_effect`, not `ID_code`.
-class side_effect_exprt:public exprt
+class side_effect_exprt : public multi_ary_exprt
```
Comment (tautschnig) [inline src/util/std_code.h]: I don't think that will work out: a `side_effect_expr_nondett` is a `side_effect_exprt` without operands.

File: src/util/std_code.h [hunk 2]
```diff
@@ -1577,7 +1578,19 @@ class side_effect_exprt:public exprt
     const irep_idt &statement,
     const typet &_type,
     const source_locationt &loc)
-    : exprt(ID_side_effect, _type)
+    : multi_ary_exprt(ID_side_effect, _type)
```
Comment (tautschnig) [inline src/util/std_code.h]: This constructor is deprecated.

**Review by tautschnig — CHANGES_REQUESTED** [review]
This is great, except for the change to the inheritance hierarchy as noted below.

File: src/util/std_code.h (hunk 1 above)
Comment (kroening) [reply src/util/std_code.h]: mutli_ary_exprt wasn't meant to guarantee that there's at least one operand -- however, reverting for now.

Comment (tautschnig) [reply src/util/std_code.h]: Ok, might be best sorted out in a separate PR. We should certainly not introduce fresh use of deprecated functions, so either a different constructor needs to be used or some of the constructors of `multi_ary_exprt` need to be de-deprecated.

File: src/util/std_code.h (hunk 2 above)
Comment (kroening) [reply src/util/std_code.h]: now using exprt

File: src/util/std_code.h [hunk 3]
```diff
@@ -1557,7 +1557,7 @@ inline const code_expressiont &to_code_expression(const codet &code)
 /// subtypes are not subtypes of \ref codet, but they inherit directly from
 /// \ref exprt. They do have a `statement` like [codets](\ref codet), but their
 /// [id()](\ref irept::id) is `ID_side_effect`, not `ID_code`.
-class side_effect_exprt:public exprt
+class side_effect_exprt : public exprt
```
Comment (tautschnig) [inline src/util/std_code.h:1560]: Nit-pick: this one and the next one below are now unrelated whitespace changes.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 044b57252672b07f16a977c5e68248d794122d0d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98134012
